File: ace.py
"""
ACE — Attention-Weighted Context Eviction

Content-aware context compression for multi-turn agentic LLM inference.

Usage:
    from ace import ace_compress, ace_evict

    # Compress a single tool-result string
    compressed = ace_compress(tool_result_text, target_ratio=0.4)

    # Evict old tool results from a messages list in-place
    saved = ace_evict(messages, budget_chars=40000, keep_recent=2)
"""

import re
from typing import List, Dict, Any


def ace_classify_line(s: str) -> float:
    """
    Score a single line for importance. Returns float in [0, 1].

    Priority:
      1.0  tool call JSON
      0.95 errors / exceptions
      0.9  file / workspace paths
      0.85 task-framing keywords
      0.7  numeric data
      0.6  shell commands
      0.5  default
      0.3  success boilerplate
      0.2  very long lines (>200 chars)
      0.1  meta-commentary
      0.0  blank lines
    """
    if not s.strip():
        return 0.0

    sl = s.lower()

    # Tool call JSON
    if '"name"' in s and ('"arguments"' in s or '"parameters"' in s):
        return 1.0

    # Errors
    error_phrases = ['error', 'failed', 'not found', 'traceback', 'exception', 'exit code']
    if any(p in sl for p in error_phrases):
        return 0.95

    # File paths
    if any(p in s for p in ['/results/', '/workspace/', '/tmp/']):
        return 0.9

    # Task keywords
    if any(k in sl for k in ['task:', 'begin', 'step ', 'verify', 'inspect']):
        return 0.85

    # Numeric data
    if re.search(r'\d+\.?\d*%|\d{4,}|bytes|rows|columns', sl):
        return 0.7

    # Shell commands
    if any(s.startswith(k) for k in ['$', '>>>', 'root@', 'agent@']) or \
       any(k in sl for k in ['wget', 'curl', 'pip', 'python3', 'apt-get', 'docker']):
        return 0.6

    # Success boilerplate
    if sl.strip() in {'done', 'ok', 'installed', 'saved', 'complete'}:
        return 0.3

    # Long verbose dump
    if len(s) > 200:
        return 0.2

    # Meta-commentary
    if any(k in sl for k in ['i executed', 'here are the results', 'tool results:']):
        return 0.1

    return 0.5


def ace_compress(content: str, target_ratio: float = 0.4) -> str:
    """
    Compress a multi-line string to ~target_ratio of its original line count.

    Always keeps first/last line. Omitted runs become a single marker line.

    Args:
        content:      Text to compress.
        target_ratio: Fraction of lines to keep (default 0.4 = 40%).

    Returns:
        Compressed string.
    """
    lines = content.split('\n')
    if len(lines) <= 3:
        return content

    scored = []
    for i, line in enumerate(lines):
        score = ace_classify_line(line)
        if i == 0 or i == len(lines) - 1:
            score = max(score, 0.7)
        scored.append((i, line, score))

    n_keep = max(2, int(len(lines) * target_ratio))
    keep_set = set(
        x[0] for x in sorted(scored, key=lambda x: -x[2])[:n_keep]
    )
    keep_set.add(0)
    keep_set.add(len(lines) - 1)

    out = []
    skipped = 0
    for i, line in enumerate(lines):
        if i in keep_set:
            if skipped > 0:
                out.append(f'  [...{skipped} lines omitted by ACE...]')
            skipped = 0
            out.append(line)
        else:
            skipped += 1

    if skipped > 0:
        out.append(f'  [...{skipped} lines omitted by ACE...]')

    return '\n'.join(out)


def ace_evict(
    messages: List[Dict[str, Any]],
    budget_chars: int = 40_000,
    keep_recent: int = 2,
    target_ratio: float = 0.4,
    min_size: int = 200,
) -> int:
    """
    Compress old tool results in a messages list until total size <= budget_chars.

    Operates on messages whose content starts with "[Tool result]: ".
    Modifies messages in-place. Skips the keep_recent most-recent tool results.

    Args:
        messages:     Anthropic-style messages list (dicts with 'role'/'content').
        budget_chars: Target max total chars.
        keep_recent:  Number of recent tool results to leave untouched.
        target_ratio: ACE compression ratio (default 0.4).
        min_size:     Don't compress messages shorter than this.

    Returns:
        Total chars removed.
    """
    def total_chars(msgs):
        return sum(len(m.get('content', '')) for m in msgs if isinstance(m.get('content'), str))

    if total_chars(messages) <= budget_chars:
        return 0

    prefix = '[Tool result]: '
    tool_indices = [
        i for i, m in enumerate(messages)
        if m.get('role') == 'user'
        and isinstance(m.get('content'), str)
        and m['content'].startswith(prefix)
    ]

    compressable = tool_indices[:len(tool_indices) - keep_recent]
    total_saved = 0

    for idx in compressable:
        if total_chars(messages) <= budget_chars:
            break
        result = messages[idx]['content'][len(prefix):]
        if len(result) <= min_size:
            continue
        if 'omitted by ACE' in result:
            continue

        compressed = ace_compress(result, target_ratio)
        saved = len(result) - len(compressed)
        if saved <= 0:
            continue

        total_saved += saved
        messages[idx]['content'] = prefix + compressed

    return total_saved

File: ace.js
/**
 * ACE — Attention-Weighted Context Eviction
 *
 * Content-aware context compression for multi-turn agentic LLM inference.
 * Replaces head/tail truncation with line-level importance scoring so that
 * tool results are compressed rather than blindly sliced.
 *
 * Usage (Node.js):
 *   const { aceCompress, aceEvict } = require('./ace');
 *
 *   // Compress a single tool-result string to ~40% of its original size,
 *   // keeping the highest-importance lines.
 *   const compressed = aceCompress(toolResultText, 0.4);
 *
 *   // Evict old tool results from a messages array in-place until the
 *   // total estimated size is under budgetChars.
 *   aceEvict(messages, { budgetChars: 40000, keepRecent: 2, targetRatio: 0.4 });
 */

'use strict';

/**
 * Score a single line for importance.
 * Returns a float in [0, 1] — higher = keep.
 *
 * Heuristic priority order:
 *  1.0  tool call JSON (name + arguments)
 *  0.95 error / exception lines
 *  0.9  file / workspace paths
 *  0.85 task-framing keywords (task:, step, verify…)
 *  0.7  numeric data (%, large numbers, units)
 *  0.6  shell commands
 *  0.5  default (unknown content)
 *  0.3  success boilerplate (done, ok, installed…)
 *  0.2  very long lines (>200 chars — usually verbose dumps)
 *  0.1  meta-commentary ("here are the results", "i executed"…)
 *  0.0  blank lines
 */
function aceClassifyLine(s) {
  const sl = s.toLowerCase();
  if (!s.trim()) return 0.0;
  // Tool call JSON
  if (s.includes('"name"') && (s.includes('"arguments"') || s.includes('"parameters"'))) return 1.0;
  // Errors
  const errorPhrases = ['error', 'failed', 'not found', 'traceback', 'exception', 'exit code'];
  if (errorPhrases.some(p => sl.includes(p))) return 0.95;
  // File paths
  if (s.includes('/results/') || s.includes('/workspace/') || s.includes('/tmp/')) return 0.9;
  // Task keywords
  if (['task:', 'begin', 'step ', 'verify', 'inspect'].some(k => sl.includes(k))) return 0.85;
  // Numeric data
  if (/\d+\.?\d*%|\d{4,}|bytes|rows|columns/.test(sl)) return 0.7;
  // Shell commands
  if (['$', '>>>', 'root@', 'agent@'].some(k => s.startsWith(k)) ||
      ['wget', 'curl', 'pip', 'python3', 'apt-get', 'docker'].some(k => sl.includes(k))) return 0.6;
  // Success boilerplate
  if (['done', 'ok', 'installed', 'saved', 'complete'].includes(sl.trim())) return 0.3;
  // Long verbose dump
  if (s.length > 200) return 0.2;
  // Meta-commentary
  if (['i executed', 'here are the results', 'tool results:'].some(k => sl.includes(k))) return 0.1;
  return 0.5;
}

/**
 * Compress a multi-line string to approximately `targetRatio` of its original
 * line count, keeping the highest-scoring lines.
 *
 * Always keeps the first and last line.
 * Omitted runs are replaced with a single "[...N lines omitted by ACE...]" marker.
 *
 * @param {string}  content      - The text to compress.
 * @param {number}  targetRatio  - Fraction of lines to keep (default 0.4).
 * @returns {string}
 */
function aceCompress(content, targetRatio = 0.4) {
  const lines = content.split('\n');
  if (lines.length <= 3) return content;

  const scored = lines.map((line, i) => {
    let score = aceClassifyLine(line);
    // Always protect first and last line
    if (i === 0 || i === lines.length - 1) score = Math.max(score, 0.7);
    return { i, line, score };
  });

  const nKeep = Math.max(2, Math.floor(lines.length * targetRatio));
  const keepSet = new Set(
    [...scored].sort((a, b) => b.score - a.score).slice(0, nKeep).map(x => x.i)
  );
  keepSet.add(0);
  keepSet.add(lines.length - 1);

  const out = [];
  let skipped = 0;
  for (let i = 0; i < lines.length; i++) {
    if (keepSet.has(i)) {
      if (skipped > 0) out.push(`  [...${skipped} lines omitted by ACE...]`);
      skipped = 0;
      out.push(lines[i]);
    } else {
      skipped++;
    }
  }
  if (skipped > 0) out.push(`  [...${skipped} lines omitted by ACE...]`);
  return out.join('\n');
}

/**
 * Evict (compress) old tool results from an Anthropic-style messages array
 * until the estimated total size is under `budgetChars`.
 *
 * Operates on messages that start with "[Tool result]: " (the format used by
 * agent-bridge.js). Skips the `keepRecent` most-recent tool results.
 * Messages are mutated in-place.
 *
 * @param {Array}   messages           - Anthropic messages array.
 * @param {Object}  opts
 * @param {number}  opts.budgetChars   - Target max total chars (default 40000).
 * @param {number}  opts.keepRecent    - Tool-result messages to leave untouched (default 2).
 * @param {number}  opts.targetRatio   - Compression ratio for aceCompress (default 0.4).
 * @param {number}  opts.minSize       - Don't compress messages shorter than this (default 200).
 * @returns {number} Total chars removed.
 */
function aceEvict(messages, opts = {}) {
  const {
    budgetChars  = 40000,
    keepRecent   = 2,
    targetRatio  = 0.4,
    minSize      = 200,
  } = opts;

  const estimateSize = (msgs) =>
    msgs.reduce((s, m) => s + (typeof m.content === 'string' ? m.content.length : 0), 0);

  if (estimateSize(messages) <= budgetChars) return 0;

  const toolIndices = messages
    .map((m, i) => (m.role === 'user' && typeof m.content === 'string'
      && m.content.startsWith('[Tool result]: ') ? i : -1))
    .filter(i => i >= 0);

  const compressable = toolIndices.slice(0, toolIndices.length - keepRecent);
  let totalSaved = 0;

  for (const idx of compressable) {
    if (estimateSize(messages) <= budgetChars) break;
    const prefix = '[Tool result]: ';
    const result = messages[idx].content.slice(prefix.length);
    if (result.length <= minSize) continue;
    if (result.includes('[...ACE') || result.includes('omitted by ACE')) continue;

    const compressed = aceCompress(result, targetRatio);
    const saved = result.length - compressed.length;
    if (saved <= 0) continue;

    totalSaved += saved;
    messages[idx].content = prefix + compressed;
  }

  return totalSaved;
}

module.exports = { aceClassifyLine, aceCompress, aceEvict };
